# Lab: Hypothesis Testing

## Group
Enter the next Python cell to add information about who you’re working with within your lab section:



In [48]:
partner1_name = "Rachael Gu"
partner1_netid = "rgu6"
partner1_major = "stat"
 
partner2_name = "Meg Li"
partner2_netid = "rmli224"
partner2_major = "stat+cs"
 
partner3_name = ""
partner3_netid = ""
partner3_major = ""

## Getting Help
Remember, there are a lot of ways to get help if you find yourself stuck:

1. In your lab section, you can ask your TA or CAs.
2. Post your question in the course Discord.
3. Office Hours:
  - In-person office hours **every Monday, Tuesday, Thursday, and Friday** from 4:00 PM - 6:00 PM in Siebel Center for Design (SCD) Rm. 0060
  - Zoom office hours **every Wednesday** from 5:00 PM - 6:00 PM and **every Thursday** from 9:00 PM - 11:00 PM (see the course website for links).

In this lab, you will:

- Perform hypothesis tests using Python.
- Create confidence intervals using Python.
- Interpret your results of the tests.

As you go through the lab:

- Make sure to run every cell with Python code and observe the output.
- Complete all puzzles and submit this lab before Monday evening at 11:59 PM.

# Setup


At the beginning of this lab, make sure you:
1. Make sure the following data sets are in **your working directory**. If not, download them from your personal GitHub repository.
- `hello.csv`
- `cs225-final-grades.csv`
- `baseball.csv`
2. Import the following libraries:
- `pandas` as `pd`
- `random`
- `matplotlib.pyplot` as `plt`
- `numpy` as `np`
- `scipy.stats`
- `from scipy import stats`
3. Read the date sets into a data frame using `pd.read.csv()`

In [49]:
# Import the libraries listed above.
import pandas as pd
import random
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats
from scipy.stats import stats

# ❓Group Discussion❓


Before we start doing any data analysis, we must determine a crucial component of the hypothesis test. During lecture, we learned about two different kinds of tests that data scientists use in hypothesis testing and confidence intervals. One is the Z-test, and the other is the t-test. Each one has its use in hypothesis testing and the creation of confidence intervals. When do you use a Z-test? When do you use a t-test?

# ❓Individual Reflection❓ 

A Z-test should be used when the standard deviation of a population is known. A T-test should be used when the sample size is small, the population is close to the normal curve, and the standard deviation of the population is unknown.

We will be using a t-test for every hypothesis test and confidence interval in this lab. While we do not necessarily have a small sample size for Puzzles 1 and 3, the sample size is not necessarily large either. We also do not know the population standard deviation, so we will lean towards using a t-test to be extra safe. A t-test is the more conservative test to use so we will go with this :)

# Puzzle 1: Sleep Among STAT 107 Students



## More Sleep Always! 

SLEEP! It's something that most of us do not get enough of. There has been extensive research done looking at the amount of sleep different people get on average. The following is an example of one of them. 

In 2013, [Gallup completed their latest survey on sleep](https://news.gallup.com/poll/166553/less-recommended-amount-sleep.aspx) that shows that the average amount of sleep Americans get is 6.8 hours ($\mu$=6.8). This value seems a bit low, right? How do you think our class average in Data Science DISCOVERY compares to this? It might be less than 6.8, or it may be more than that. Let's find out if the ideal hours of sleep STAT107 students would get is significantly different from the average amount of sleep that all Americans get reported by Gallup. For the sake of this lab, assume our survey data is a random sample from the population (Americans).

Run a **hypothesis test** to determine if the STAT107 data is significantly different from the population/Gallup poll. 

You need to look at the `hello.csv` data set that gives you the ideal amount of sleep STAT107 students get each week.

As always, start with uploading the data set and import the relevant libraries to your notebook.


In [50]:
#Import the hello data set.
hello = pd.read_csv("hello.csv")

In [51]:
#TEST CASE
## == TEST CASES for Puzzle 1 Data ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

assert(len(hello) == 284), "The dataframe does not appear to have correct dataset loaded."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")
print()

🎉 All tests passed! 🎉



## Part 1: Null and Alternative Hypothesis

In class, we learned that the first step when doing any hypothesis test is to state the **Null Hypothesis** and **Alternative Hypothesis**. 

**Null Hypothesis:** The null hypothesis says that any difference between our sample and the population is small and therefore due to chance. 

**Alternative Hypothesis:** The alternative hypothesis says that the differences we observe are not just due to chance. 

Remember, when you are stating your hypotheses, they always include the **population parameter** $\mu$. This is the population mean! Do not use the sample statistic (the sample mean) $\bar{x}$.

To type $\mu$ in Markdown, type, `$\mu$`.

To type $\neq$ in Markdown, type, `$\neq$`

## ❓Individual Reflection❓ 

State your two hypotheses below by editing the next cell! Here, we want to use a two sided alternative hypothesis since we aren't specifying a direction.

### **Null Hypothesis**

$H_0$: hours of sleep for STAT107 students = $\mu$


### **Alternative Hypothesis**

$H_a$: hours of sleep for STAT107 students $\neq$ $\mu$


## Part 2: Doing the Hypothesis Test


Now that we have stated our null and alternate hypotheses, it is time to see which one is supported by our data. Before we can go and do any coding, we need to look at some summary statistics about our data. These will not necessarily tell us which of our hypotheses is supported by the data but will be crucial when performing the test and constructing the confidence interval.

## Part 2a: Summary Statistics


Calculate the sample mean $\bar{x}$, standard deviation $\hat{\sigma}$, sample size “n”, and the degrees of freedom “df” for the hours of sleep for STAT107 students in the `hello.csv` data set. Have the cell print out all four values.

In [52]:
hrs_sleep_mean = hello["Sleep Hours"].mean()
hrs_sleep_std = hello["Sleep Hours"].std()
sample_size = len(hello)
df = sample_size - 1
print("The average amount of sleep that students get is " + str(hrs_sleep_mean) + " hours.")
print("The standard deviation of sleep that students get is " + str(hrs_sleep_std) + " hours.")
print("The sample size is " + str(sample_size) + " students.")
print("The degrees of freedom is " + str(df) + ".")

The average amount of sleep that students get is 7.125 hours.
The standard deviation of sleep that students get is 1.0965936657871607 hours.
The sample size is 284 students.
The degrees of freedom is 283.


In [53]:
#TEST CASE
### == TEST CASES for Puzzle 1.2a ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(hrs_sleep_mean, 7.125, abs_tol= 0.05)), "The mean does not appear to seem correct. Make sure you are taking the mean of the correct variable."
assert(math.isclose(hrs_sleep_std, 1.0965936657871607, abs_tol= 0.05)), "The standard deviation does not appear to seem correct. Make sure you are taking the std of the correct variable."
assert(math.isclose(sample_size, 284)), "The sample size does not appear to seem correct. Make sure you are taking the sample size of the correct dataframe."
assert(math.isclose(df, 283)), "The df does not appear to seem correct. Make sure you are taking the df of the correct dataframe."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")
print()

🎉 All tests passed! 🎉



## Part 2b: t-Test Statistic, t-Critical Value, and p-value

Calculate the t-test statistic, t-critical value, and p-value. This will tell us whether we will reject or fail to reject our null hypothesis.  

The test statstic is t which you will find from the formula for t. Check your notes if you forgot this :)

The critical value is the t-value when there is 97.5% of the t curve to the left (or 5% in each tail). We are using 5% as alpha. The critical value also depends on the degrees of freedom!

**Hint 1:** Remember to use the df that you calculated above instead of n when calculating the t-Critical Value and p-value.

**Hint 2:** Remember to use $\frac{\alpha}{2}$ when calculating the t-Critical Value.  Also, make sure to calculate your p-value based on your alternative hypothesis! Is this a one sided test or a 2 sided test?

In [54]:
from scipy.stats import norm
SD_plus = hrs_sleep_std * ((sample_size / df)**(1/2))
standard_error = SD_plus / (sample_size**(1/2))
t_stat = (hrs_sleep_mean - 6.8) / standard_error
critical_value = scipy.stats.t.ppf(0.975, df)
p_value = 2 * scipy.stats.t.sf(t_stat, df)
print("The t-test statistic is " + str(t_stat) + ".")
print("The t-critical value is " + str(critical_value) + ".")
print("The p-value for the hypothesis test is " + str(p_value) + ".")

The t-test statistic is 4.985753993467715.
The t-critical value is 1.9683819234318902.
The p-value for the hypothesis test is 1.0775989739128053e-06.


In [55]:
#TEST CASE
### == TEST CASES for Puzzle 1.2b ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(t_stat, 4.994554978240841, abs_tol= 0.05)), "Your t statistic does not seem correct. Make sure you are using the correct formula"
assert(math.isclose(critical_value, 1.9683819234318902, abs_tol= 0.05)), "Your critical value does not seem to be correct. Did you use α instead of α/2?"
assert(math.isclose(p_value, 1.0334128940202397e-06, abs_tol= 0.0000001)), "Your p_value does not seem to be correct. Make sure you are inputing the right values."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 2c: Confidence Interval


You now know whether you reject or fail to reject your null hypothesis since we calculated the p-value. However, we can still get some more information by creating a confidence interval around the sample mean. For this part, construct a 95% confidence interval for the sample mean of the hours of sleep for STAT107 students. 

A confidence interval with t is the same format as a confidence interval for z. See below:

**($\bar{x}$ - $t_{\frac{\alpha}{2}, df}$(SE) , $\bar{x}$ + $t_{\frac{\alpha}{2}, df}$(SE))**

Where $t_{\frac{\alpha}{2}, df}$ is the t-critical value (critical_value) that you calculated in part b.

In [58]:
lower_bound = hrs_sleep_mean - (1.96 * standard_error)
upper_bound = hrs_sleep_mean + (1.96 * standard_error)
print("The 95% CI is:(" + str(lower_bound) + ", " + str(upper_bound) + ").")

The 95% CI is:(6.997235974571832, 7.252764025428168).


In [59]:
#TEST CASE
### == TEST CASES for Puzzle 1.2c ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(lower_bound, 6.9956926772702355, abs_tol= 0.05)), "Your lower bound does not seem correct. Make sure your formula is correct."
assert(math.isclose(upper_bound, 7.2552436523177795, abs_tol= 0.05)), "Your upper bound does not seem to be correct. Make sure your formula is correct."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


# Part 3: Interpretation


As a data scientist, it is crucial to explain your results, so anyone (even someone who has no background in statistics) can understand. If we just said, "Here is the p-value.", they may be confused. We need to be more specific about our conclusion.

## Part 3a: The Result


In the following cell, delete the statements you do not believe are correct and keep the one you think is.

In [60]:
# Remove two of the following lines to leave only the correct answer:
Conclusion = "There is enough evidence to reject the null hypothesis. Thus, we reject the null hypothesis."

In [61]:
#TEST CASE
### == TEST CASES for Puzzle 1.3a ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert(Conclusion.split()[2][3::-1] == "uone"), "Your conclusion does not seem correct. Think about it some more!"

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## ❓Part 3b: Group Discussion❓

We have now done our hypothesis test and discovered the result, but there is still one more crucial thing to do. We must be able to explain the results of the test to somebody who does not know statistics. In the future, your occupation will require this as a skill. With your group, discuss the results of what your hypothesis test means, and interpret its practical significance if there is any. Use **NON-STATISTICAL** terminology here. In other words, describe the conclusion of the t-test and what it means in the context of the problem. Don't just say "reject or do not reject Ho."

## ❓Individual Reflection❓

Based on this analysis, we can conclude that the average hours of sleep for STAT107 students is not equal to 6.8 hours. 

# Puzzle 2: GPA Visualization

Let's do another t test to make sure you have the hang of it! Let's say you have to sign up for CS 225 (DATA STRUCTURES) next semester. Due to your apartment’s terrible internet, you were unable to sign up when registration opened. By the time your internet comes back, the only section left is with Dr. Fagen-Ulmschneider. You decide to look up how difficult your semester will be, and you find something useful. Dr. Fagen-Ulmschneider has a [Grade disparity between sections at UIUC](http://waf.cs.illinois.edu/discovery/grade_disparity_between_sections_at_uiuc/) visualization. First, find the average GPA and closest letter grade to that Dr. Fagen-Ulmschneider gave his CS 225 students.

**Hint:** This will be what you are testing ($\mu$ = “average GPA”).



## ❓Individual Reflection❓
Replace the first ? with the average GPA (use Wade's visualization to see what it is!).  Replace the second ? with the letter grade that corresponds to that GPA. You do not need to do any calculations, just look at the GPA visualization and the translation to letter grade below.

The average GPA is 3.22

The average letter grade is B

You want to double-check if Dr. Fagen-Ulmschneider was completely honest with his grades. You randomly survey 107 students who have taken CS 225. The final grades the students gave you are out of 1100 points. You have recorded the 107 student grades in `cs225-final-grades.csv`. *(These are 107 randomly chosen grades from CS 225 from Fall 2018.)*

You find that the translation between points and grades is as follows:


| Points | Grade | GPA |
| ------ | ----- | --- |
| 1070 - 1100 | A+ | 4.0 |
|  930 - 1070 | A  | 4.0 |
|  900 - 930 | A-  | 3.67 |
|  870 - 900 | B+  | 3.33 |
|  830 - 870 | B   | 3.0 |
|  800 - 830 | B-  | 2.67 |
|  770 - 800 | C+  | 2.33 |
|  730 - 770 | C   | 2.0 |
|  700 - 730 | C-  | 1.67 |
|  670 - 700 | D+  | 1.33 |
|  630 - 670 | D   | 1.0 |
|  600 - 630 | D-  | 0.67 |
|    0 - 600 | F  | 0 |


**Hint:** You will need to process `cs225-final-grades.csv`, understand, and gather data from the GPA visualization before you will be able to run the test. Use `pandas` to help you process this data.

In [62]:
#Import the Data Set
cs225 = pd.read_csv("cs225-final-grades.csv")
cs225

CS 225 Final Grade
0            658.204335
1           1068.696078
2           1040.655631
3           1047.147059
4            763.440841
..                  ...
102          954.931372
103          899.918937
104          765.717112
105          328.562675
106         1048.897892

[107 rows x 1 columns]

In [63]:
#TEST CASE
## == TEST CASES for Puzzle 2 Data ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

assert(len(cs225) == 107), "The dataframe does not appear to have correct dataset loaded."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")
print()

🎉 All tests passed! 🎉



## Part 1: Null and Alternative Hypothesis

Just like in Puzzle 1, start with stating your Null and Alternative Hypothesis. Remember that when you are stating your hypotheses, they always include the **population parameters** such as $\mu$, not sample statistics such as $\bar{x}$. Again, we are going to use a **two sided** alternative hypothesis test!

To type $\mu$ in Markdown, type, `$\mu$`.

To type $\neq$ in Markdown, type, `$\neq$`

## ❓Individual Reflection❓ 

State your two hypotheses by editing the cell below:

### Null Hypothesis

$H_0$: average GPA in CS 225 = $\mu$

### Alternative Hypothesis

$H_a$: average GPA in CS 225 $\neq$ $\mu$

## Part 2: Doing the Hypothesis Test
Once again, look at some summary statistics about our data.

## Part 2a: Summary Statistics

Calculate the sample mean $\bar{x}$, standard deviation $\hat{\sigma}$, sample size “n”, and the degrees of freedom “df” for students’ GPA for CS225 in the `cs225-final-grades.csv` data set. Have the cell print out all four values.

In [65]:
cs225_sample_mean = cs225["CS 225 Final Grade"].mean()
cs225_sample_std = cs225["CS 225 Final Grade"].std()
cs225_sample_size = len(cs225)
cs225_df = cs225_sample_size - 1
print("The average amount of points that a student gets in CS225 is " + str(cs225_sample_mean) + " points.")
print("The standard deviation of points that a student gets in CS225 is " + str(cs225_sample_std) + " points.")
print("The sample size is " + str(cs225_sample_size) + " students.")
print("The degrees of freedom is " + str(cs225_df) + ".")

The average amount of points that a student gets in CS225 is 915.3368698934578 points.
The standard deviation of points that a student gets in CS225 is 182.19112009595577 points.
The sample size is 107 students.
The degrees of freedom is 106.


In [66]:
#TEST CASE
### == TEST CASES for Puzzle 2.2a ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(cs225_sample_mean, 915.3368698934579, abs_tol= 0.05)), "The mean does not appear to seem correct. Make sure you are taking the mean of the correct variable."
assert(math.isclose(cs225_sample_std, 182.1911200959558, abs_tol= 0.05)), "The standard deviation does not appear to seem correct. Make sure you are taking the std of the correct variable."
assert(math.isclose(cs225_sample_size, 107)), "The sample size does not appear to seem correct. Make sure you are taking the sample size of the correct dataframe."
assert(math.isclose(cs225_df, 106)), "The df does not appear to seem correct. Make sure you are taking the df of the correct dataframe."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")
print()

🎉 All tests passed! 🎉



## Part 2b: t-Test Statistic, t-Critical Value, and p-value

Calculate the t-test statistic, t-critical value, and p-value just like we did for Puzzle 1.


**Hint 1:** Remember to use the df that you calculated above instead of n when calculating the t-Critical Value and p-value. Also, take note if your p-value should be two sided or one sided!

**Hint 2:** Remember to use $\frac{\alpha}{2}$ when calculating the t-Critical Value.

**Hint 3:** Translate your $\bar{x}$ and $\sigma$ into GPA points. If $\bar{x}$ > 930 points, then set $\bar{x}$ = 4.00. If $\bar{x}$ is between 900 and 930, then set $\bar{x}$ = 3.67. Etc. Don't forget to do this for the calculations below!

When doing your calculations below, use $\sigma$ = 1.67.  

In [67]:
SD_plus = ((cs225_sample_size / cs225_df)**0.5) * 1.67
standard_error = SD_plus / (cs225_sample_size**0.5)
cs225_t_stat = (3.67-3.22) / standard_error
cs225_critical_value = scipy.stats.t.ppf(0.975, cs225_df)
cs225_p_value = 2 * scipy.stats.t.sf(cs225_t_stat, cs225_df)
print("The t-test statistic is " + str(cs225_t_stat) + ".")
print("The t-critical value is " + str(cs225_critical_value) + ".")
print("The p-value for the hypothesis test is " + str(cs225_p_value) + ".")

The t-test statistic is 2.7742715948767347.
The t-critical value is 1.9825972617102907.
The p-value for the hypothesis test is 0.006539701273862873.


In [68]:
#TEST CASE
### == TEST CASES for Puzzle 2.2b ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(cs225_t_stat, 2.7873270627274658, abs_tol= 0.05)), "Your t statistic does not seem correct. Make sure you are using the correct formula."
assert(math.isclose(cs225_critical_value, 1.9825972617102907, abs_tol= 0.05)), "Your critical value does not seem to be correct. Did you use α instead of α/2?"
assert(math.isclose(cs225_p_value, 0.006299433880445842, abs_tol= 0.0005)), "Your p_value does not seem to be correct. Make sure you are inputing the right values."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 2c: Confidence Interval

Construct a 95% confidence interval for the sample mean of the student’s GPA for CS225. Use the following formula. 

**($\bar{x}$ - $t_{\frac{\alpha}{2}, df}$(SE) , $\bar{x}$ + $t_{\frac{\alpha}{2}, df}$(SE))**

Where $t_{\frac{\alpha}{2}, df}$ is the t-critical value that you calculated in part b.

In [69]:
lower_bound = 3.67 - (1.96 * standard_error)
upper_bound = 3.67 + (1.96 * standard_error)
print("The 95% CI is:(" + str(lower_bound) + ", " + str(upper_bound) + ").")

The 95% CI is:(3.35207871153322, 3.98792128846678).


In [70]:
#TEST CASE
### == TEST CASES for Puzzle 2.2c ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

assert(math.isclose(lower_bound, 3.349919597631782, abs_tol= 0.05)), "Your lower bound does not seem correct. Make sure your formula is correct."
assert(math.isclose(upper_bound, 3.9900804023682177, abs_tol= 0.05)), "Your upper bound does not seem to be correct. Make sure your formula is correct."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 3: Interpretation

As with the last puzzle, the interpretation of the results for the hypothesis test is important.

## Part 3a: The Result

In the following cell, delete the statements that you do not believe are true and keep the one you believe is correct.

In [71]:
# Remove two of the following lines to leave only the correct answer:
Conclusion = "There is enough evidence to reject the null hypothesis. Thus, we reject the null hypothesis."

In [72]:
#TEST CASE
### == TEST CASES for Puzzle 2.3a ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert(Conclusion.split()[2][3::-1] == "uone"), "Your conclusion does not seem correct. Think about more!"

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## ❓Part 3b: Group Discussion❓

When you get back to campus and sit in your first lecture, you see a friend of yours. They are concerned about taking this course due to reading horror stories about it on Reddit. They did not do any of the research that you did. They also have never taken a statistics course before. What will you end up telling them? With your group, discuss the results of what your hypothesis test means, and interpret its practical significance if there is any. Use **NON-STATISTICAL** terminology here.

Based on the analysis, the average GPA for STAT107 students is not 3.22. The average GPA for STAT107 students is likely to be from 3.35 to 3.98. 

# Puzzle 3: Baseball Analysis


One final t test!!  Over the summer, you get an internship for a high school summer baseball league. In this league, there are four different levels for each age group. The levels (in descending order) are the Invite, the Elite, the Select, and the Enlist. 

As you start the last few days of your internship, your supervisor asks to see you. He asks a favor of you before you leave. Each year, the winner of each level below Invite may petition the league to move up to the next highest level for next summer. All of the petitions but one have been approved or denied. Their statistics intern left already and forgot to look over the last petition. They are on a family vacation before the semester starts, so your supervisor cannot get ahold of them. You had once mentioned you took a statistics course, so your supervisor asks for help. He has gone through the petition. However, before making the final decision, he wants to know if the team’s batting average is statistically different from the Select’s overall. He knows a hypothesis test is required, but does not remember how to perform it. **Fortunately, you still have VS Code and Python installed on your laptop ;)**

In [73]:
#Import the Baseball Data Set
bb = pd.read_csv("baseball.csv")
bb

Unnamed: 0  Number  GP  PA  AB   H  H_1B  H_2B  H_3B  HR  ...  DP  TP  \
0            1       1  29  93  71  20    12     7     1   0  ...   1   0   
1            2       3  29  96  73  29    27     2     0   0  ...   3   0   
2            3       4  29  84  77  30    25     5     0   0  ...   1   0   
3            4       9  29  98  74  22    16     4     2   0  ...   0   0   
4            5      17  21  39  22   4     3     1     0   0  ...   0   0   
5            6      21  29  77  56  17    16     1     0   0  ...   2   0   
6            7      22  27  35  25   2     2     0     0   0  ...   0   0   
7            8      23  27  76  63  20    13     4     3   0  ...   2   0   
8            9      24  27  68  53   9     7     1     1   0  ...   0   0   
9           10      27   1   1   0   0     0     0     0   0  ...   0   0   
10          11      27  25  68  55  15     8     7     0   0  ...   2   0   
11          12      29  25  62  54  13    13     0     0   0  ...   5   0   
12          13      32  29  91  71  18    17     1     0   0  ...   2   0   

     FPCT   cINN  cPB  cSB  cCS  cCSprct  cPIK  cCI  
0   0.957  134.1    2   20    5    0.200     3    0  
1   0.938    0.0    0    0    0    0.000     0    0  
2   0.909    0.0    0    0    0    0.000     0    0  
3   0.941    0.0    0    0    0    0.000     0    0  
4   0.915   41.2    5   11    1    0.083     0    0  
5   0.939    0.0    0    0    0    0.000     0    0  
6   0.938    4.0    3    2    0    0.000     0    0  
7   0.886    0.0    0    0    0    0.000     0    0  
8   1.000    0.0    0    0    0    0.000     0    0  
9   0.000    0.0    0    0    0    0.000     0    0  
10  0.871    0.0    0    0    0    0.000     0    0  
11  0.901    0.0    0    0    0    0.000     0    0  
12  0.926    0.0    0    0    0    0.000     0    0  

[13 rows x 130 columns]

In [74]:
#TEST CASE
## == TEST CASES for Puzzle 3 Data ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

assert(len(bb) == 13), "The dataframe does not appear to have correct dataset loaded."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")
print()

🎉 All tests passed! 🎉



## Part 1: Null and Alternative Hypothesis

## ❓Individual Reflection❓ 

State your Null and Alternative Hypothesis. Again, use a two sided alternative.  Edit the next cell to do this!

To type $\mu$ in Markdown, type, `$\mu$`.

To type $\neq$ in Markdown, type, `$\neq$`

**Hint:** The mean batting average for the Select teams over the summer was 0.212.

### Null Hypothesis

$H_0$: the mean batting average for the Enlist teams = $\mu$

### Alternative Hypothesis

$H_a$: the mean batting average for the Enlist teams $\neq$ $\mu$

## Part 2: Doing the Hypothesis Test


Once again, look at some summary statistics about our data.


## Part 2a: Summary Statistics


Calculate the sample mean $\bar{x}$, standard deviation $\hat{\sigma}$, sample size “n”, and the degrees of freedom “df” for the team’s batting average last summer in the `baseball.csv` data set. Have the cell print out all four values.

**NOTE 1:** You **CAN NOT** calculate the mean batting average for a team by adding up all the batting averages and dividing by n. To find the mean batting average $\bar{x}$, use the formula below instead.

**NOTE 2:** With the sample mean being calculated differently, the standard deviation will have to be manually calculated as well. If you use the standard deviation function in Python, it will use the incorrect sample mean instead. To find the standard deviation, use the formula below instead.


$\bar{x}$ = $\frac{\sum_{i=1}^{n} H_{i}}{\sum_{i=1}^{n} AB_{i}}$

Or, take the total number of hits (H) and divide it by the total number of at-bats (AB).

SD = $\sqrt{\frac{1}{n-1} \sum_{i=1}^{n} (x_{i}-\bar{x})^{2}}$

Or, take the observed values and subtract the sample mean, square them, sum them all up, divide them by n-1 (the degrees of freedom), and take the square root.

In [75]:
#Python Code
values = ['AVG']

# use the formula above to find the sample mean
bb_sample_mean = bb["H"].sum() / bb["AB"].sum()

bb_sample_size = len(bb)
bb_df = len(bb) - 1

# this creates a dataframe with the mean variance values
bb_sample_var = (bb[values] - bb_sample_mean)**2

# use the formula above to find the sample SD
bb_sample_std = ((bb_sample_var.sum() / bb_df)**0.5)

print("The average batting average of the team is " + str(bb_sample_mean) + " .")
print("The standard deviation of the batting average of the team is " + str(bb_sample_std) + " points.")
print("The sample size is " + str(bb_sample_size) + " students.")
print("The degrees of freedom is " + str(bb_df) + ".")


The average batting average of the team is 0.28674351585014407 .
The standard deviation of the batting average of the team is AVG    0.121476
dtype: float64 points.
The sample size is 13 students.
The degrees of freedom is 12.


In [76]:
#TEST CASE
### == TEST CASES for Puzzle 3.2a ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(bb_sample_mean, 0.28674351585014407, abs_tol= 0.05)), "The mean does not appear to seem correct. Make sure you are taking the mean of the correct variable."
assert(math.isclose(bb_sample_std, 0.12147626669385032, abs_tol= 0.05)), "The standard deviation does not appear to seem correct. Make sure you are taking the std of the correct variable."
assert(math.isclose(bb_sample_size, 13)), "The sample size does not appear to seem correct. Make sure you are taking the sample size of the correct dataframe."
assert(math.isclose(bb_df, 12)), "The df does not appear to seem correct. Make sure you are taking the df of the correct dataframe."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")
print()

🎉 All tests passed! 🎉



## Part 2b: t-Test Statistic, t-Critical Value, and p-value

Calculate the t-test statistic, t-critical value, and p-value.

In [81]:
standard_error = bb_sample_std / np.sqrt(bb_sample_size)
bb_t_stat = (bb_sample_mean - 0.212) / standard_error
bb_critical_value = scipy.stats.t.ppf(0.975, bb_df)
bb_p_value = 2 * scipy.stats.t.sf(bb_t_stat, bb_df)
print("The t-test statistic is " + str(bb_t_stat) + ".")
print("The t-critical value is " + str(bb_critical_value) + ".")
print("The p-value for the hypothesis test is " + str(bb_p_value) + ".")

The t-test statistic is AVG    2.218471
dtype: float64.
The t-critical value is 2.1788128296634177.
The p-value for the hypothesis test is [0.04656628].


In [82]:
#TEST CASE
### == TEST CASES for Puzzle 3.2b ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

import math
assert(math.isclose(bb_t_stat, 2.218471, abs_tol= 0.05)), "You t statistics does not seem correct. Make sure you are using the correct formula."
assert(math.isclose(bb_critical_value, 2.1788128296634177, abs_tol= 0.05)), "Your critical value does not seem to be correct. Did you use α instead of α/2?"
assert(math.isclose(bb_p_value, 0.04656628335554172, abs_tol= 0.005)), "Your p_value does not seem to be correct. Make sure you are inputing the right values."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 2c: Confidence Interval

Construct a 95% confidence interval for the sample mean of the team’s batting average. Use the following formula!

**($\bar{x}$ - $t_{\frac{\alpha}{2}, df}$(SE) , $\bar{x}$ + $t_{\frac{\alpha}{2}, df}$(SE))**

Where $t_{\frac{\alpha}{2}, df}$ is the t-critical value that you calculated in part b.

In [83]:
lower_bound = 0.286 - (1.96 * standard_error)
upper_bound = 0.286 + (1.96 * standard_error)
print("The 95% CI is:(" + str(lower_bound) + ", " + str(upper_bound) + ").")

The 95% CI is:(AVG    0.219965
dtype: float64, AVG    0.352035
dtype: float64).


In [84]:
#TEST CASE
### == TEST CASES for Puzzle 3.2c ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

assert(math.isclose(lower_bound, 0.21333614256625816, abs_tol= 0.05)), "Your lower bound does not seem correct. Make sure your formula is correct."
assert(math.isclose(upper_bound, 0.36015088913403, abs_tol= 0.05)), "Your upper bound does not seem to be correct. Make sure your formula is correct."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 3: Interpretation

As with the last two puzzles, the interpretation of the results for the hypothesis test is important.

## Part 3a: The Result

In the following cell, delete the statements that you do not believe are true and keep the one you believe is correct.

In [85]:
# Remove two of the following lines to leave only the correct answer:

Conclusion = "There is enough evidence to reject the null hypothesis. Thus, we reject the null hypothesis."

In [86]:
#TEST CASE
### == TEST CASES for Puzzle 3.3a ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert(Conclusion.split()[2][3::-1] == "uone"), "Your conclusion does not seem correct. Think about it some more!"

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## ❓Part 3b: Group Discussion❓

Now that you have completed your hypothesis test, what will you tell your supervisor? With your group, discuss the results of what your hypothesis test means, and interpret its practical significance if there is any. Use **NON-STATISTICAL** terminology here.

❓Individual Reflection❓ 

From this analysis, the mean batting average for Enlist's teams is not the same as the mean batting average for Select's teams. The mean batting average for Enlist teams is around 0.22 and 0.35.

# Submit Your Work!

You are now done with the lab.

You need to do two more things before you can go back to doing other homework:

1. Save your work.  To do this, create a **notebook checkpoint** by using the menu within the notebook. Go to **File -> Save All**.

2. After you have saved it, exit this notebook and return to the Data Science Discovery page for the Git commands to submit this lab.